### PR TITLE
1/n inter batch parallelism

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduce the first component from https://github.com/PyTorchLightning/pytorch-lightning/pull/8959

Parts #8867
Parts #8316

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/fetching.py [hunk 1]
```diff
@@ -35,34 +38,60 @@
 class AbstractDataFetcher(ABC):
 
     """
-    This class is used to control batch fetching flow.
+    This based class should be used to implement a fault tolerant ``DataFetcher``.
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/fetching.py:41]:
```suggestion
    This base class should be used to implement a fault tolerant ``DataFetcher``.
```

File: pytorch_lightning/utilities/fetching.py [hunk 2]
```diff
@@ -35,34 +38,60 @@
...
+    It is required to override the ``fetching_function`` with fetching logic.
+    Example::
+        class SimpleDataFetcher(AbstractDataFetcher):
```
Comment (ethanwharris) [inline pytorch_lightning/utilities/fetching.py:44]: Need the blank lines for docs to render properly
```suggestion
    It is required to override the ``fetching_function`` with fetching logic.
    
    Example::
    
        class SimpleDataFetcher(AbstractDataFetcher):
```

**Review by ethanwharris — APPROVED** [review]
LGTM, small comments

File: pytorch_lightning/utilities/fetching.py [hunk 3]
```diff
@@ -35,34 +38,60 @@
...
+        stage: Optional[str] = None,
+        batch_to_device: Optional[Callable] = None,
+        profiler: "Optional[pl.profiler.base.BaseProfiler]" = None,
+    ) -> None:
         self._add_capture_metadata_collate(dataloader)
+
         self.dataloader = dataloader
+        self.stage = stage
+        self.batch_to_device = batch_to_device
+        self.profiler = profiler
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py:93]: are these not available at the time of creating the fetcher?

File: pytorch_lightning/loops/processors/iterator_batch_processor.py [hunk 4]
```diff
@@ -114,14 +108,33 @@ def get_active_optimizers(self, batch_idx: Optional[int] = None) -> List[Tuple[i
         """
         return list(enumerate(self.trainer.optimizers))
 
+    def _build_kwargs(self, dataloader_iter: Iterator, batch_idx: int) -> Dict[str, Any]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:111]: shouldn't this go to the bottom to the protected section?

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 5]
```diff
@@ -26,7 +26,7 @@ class DataConnector:
     def __init__(self, trainer: "pl.Trainer", multiple_trainloader_mode: str = "max_size_cycle"):
         self.trainer = trainer
         self.multiple_trainloader_mode = multiple_trainloader_mode
-        self.data_fetcher: Optional[DataFetcher] = None
+        self.data_fetcher: Optional[DataFetcher] = DataFetcher()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/data_connector.py:29]:
```suggestion
        self.data_fetcher: DataFetcher = DataFetcher()
```

File: pytorch_lightning/utilities/fetching.py [hunk 6]
```diff
@@ -35,34 +38,60 @@
 class AbstractDataFetcher(ABC):
 
     """
-    This class is used to control batch fetching flow.
+    This based class should be used to implement a fault tolerant ``DataFetcher``.
+    It is required to override the ``fetching_function`` with fetching logic.
+    Example::
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py]:
```suggestion
    It is required to override the ``fetching_function`` with fetching logic.

    Example::
```

File: pytorch_lightning/utilities/fetching.py [hunk 7]
```diff
@@ -35,34 +38,60 @@
...
+                    except StopIteration:
+                        yield None, True
     """
 
     @abstractmethod
     def fetching_function(self) -> Generator:
-        pass
+        """Override with your own fetching logic."""
 
     def __init__(
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py:59]: why is init below the fetching function 😄

File: pytorch_lightning/utilities/fetching.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/fetching.py:93]: what about creating the fetcher in on_trainer_init?

File: pytorch_lightning/utilities/fetching.py [hunk 8]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+        yield batch, is_last
+
+    def move_data_to_device(self, batch: Any) -> Any:
+        if self.batch_to_device:
+            with self.apply_profiler(f"move_{self.stage}_batch_to_device"):
+                batch = self.batch_to_device(batch)
+        return batch
+
+
+class InterBatchParallelismDataFetcher(DataFetcher):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py:308]:
```suggestion
class InterBatchParallelDataFetcher(DataFetcher):
```

File: pytorch_lightning/utilities/fetching.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:41]: Done !

File: pytorch_lightning/utilities/fetching.py [hunk 9]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+
+    """
+    This class implements inter-batch parallelism, which aims at hiding the latency of host-to-device copy
+    of input batches behind computationally intensive operations.
+
+    Without parallelization:
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:                          [HtoD][forward][backward]
+    batch 2:                                                   [HtoD][forward][backward]
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py]: probably has to be formatted as a code block because the character spacing will not be mono, so this illustration will not work.

File: pytorch_lightning/utilities/fetching.py [hunk 10]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+
+class DataLoaderIterDataFetcher(AbstractDataFetcher):
+
+    """
+    This class is used to return directly the `dataloader_iter` to the ``LightningModule`` training_step
+    for users to implement their own pre-fetching logic.
+    This feature can be activated as follows:
+
+    Example::
+        Class MyModel(LightningModule):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py]:
```suggestion
    Example::

        Class MyModel(LightningModule):
```

File: pytorch_lightning/utilities/fetching.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: tests/utilities/test_fetching.py [hunk 11]
```diff
@@ -84,3 +95,148 @@ def test_misconfiguration_error():
...
+
+    def val_dataloader(self):
+        return DataLoader(RandomIndicesDataset(), batch_size=4)
+
+    def test_dataloader(self):
+        return DataLoader(RandomIndicesDataset(), batch_size=4)
+
+
+@RunIf(min_gpus=1, min_torch="1.8.0")
+def test_trainer_num_prefetch_batches(tmpdir):
```
Comment (awaelchli) [inline tests/utilities/test_fetching.py:181]: @tchaton how expensive are these tests? I see we are waiting to simulate a heavy cuda call?

File: pytorch_lightning/utilities/fetching.py [hunk 12]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+            def __init__(self):
+                self.automatic_optimization = False
+            def training_step(self, dataloader_iter: Iterator, batch_idx: int) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/fetching.py]:
```suggestion
            def __init__(self):
                self.automatic_optimization = False
                
            def training_step(self, dataloader_iter: Iterator, batch_idx: int) -> None:
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 13]
```diff
@@ -61,8 +61,11 @@ def on_trainer_init(
         self.trainer._is_data_prepared = False
 
     def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
-        self.data_fetcher = DataFetcher()
-        self.data_fetcher.setup(train_dataloader)
+        # FIXME: Temporary hack
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:64]: why is this temporary? Do you intend to fix it here?

File: pytorch_lightning/trainer/trainer.py [hunk 14]
```diff
@@ -924,6 +925,8 @@ def _maybe_switch_to_iterator_batch_processor(self, model: "pl.LightningModule")
             )
             batch_loop = IteratorBatchProcessor(self, model)
             self.fit_loop.epoch_loop.connect(batch_loop)
+            # FIXME: Move this logic to data_connector after removing `IteratorBatchProcessor`
```
Comment (justusschock) [inline pytorch_lightning/trainer/trainer.py:928]: Will you do this in this PR or in some follow-ups?

File: pytorch_lightning/utilities/fetching.py (hunk 7 above)
Comment (justusschock) [reply pytorch_lightning/utilities/fetching.py:59]: I actually think it is good that way, since the fetching function is the abstract one to be overwritten

File: pytorch_lightning/utilities/fetching.py [hunk 15]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+    Args:
+        prefetch_batches: Number of batches to be pre-fetched. Lightning will pre-fetch
+            at least 1 batch for tracking the latest batch.
+        store_on_gpu: Whether to store the pre-fetched batches on device.
     """
 
+    def __init__(
+        self,
+        prefetch_batches: int = 0,
+        store_on_gpu: bool = False,
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/utilities/fetching.py:222]: This mainly is since we want to overlay the host-to-gpu copy here which requires some additional cuda logic (mainly streams).

But I agree, what about other accelerators?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 16]
```diff
@@ -158,6 +160,7 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
             self.update_lr_schedulers("epoch", update_plateau_schedulers=False)
 
         batch_end_outputs = [opt_idx_out for opt_idx_out in batch_output.training_step_output if len(opt_idx_out)]
+
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:163]: Done !

File: pytorch_lightning/utilities/fetching.py [hunk 17]
```diff
@@ -35,34 +38,60 @@
...
+    This based class should be used to implement a fault tolerant ``DataFetcher``.
+    It is required to override the ``fetching_function`` with fetching logic.
+    Example::
+        class SimpleDataFetcher(AbstractDataFetcher):
+            def fetching_function(self):
+                while True:
+                    try:
+                        yield next(self.dataloader_iter), False
+                    except StopIteration:
+                        yield None, True
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:222]: changed to `store_on_device`.

File: pytorch_lightning/utilities/fetching.py [hunk 18]
```diff
@@ -169,40 +198,207 @@ def reset(self) -> None:
...
+                    self.wait()
 
-    def _consume_prefetched_batches(self) -> Generator:
-        self.done = True
-        while self.batches:
-            if len(self.batches) == 1:
-                yield self.batches.pop(0), True
-            else:
-                yield self.batches.pop(0), False
+                    yield (self.move_data_to_device(yield_batch) if not self.store_on_gpu else yield_batch, False)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:257]: Yes, this is actually accelerator.batch_to_device transform with LightningModule hooks.

File: pytorch_lightning/utilities/fetching.py [hunk 19]
```diff
@@ -35,34 +38,60 @@
...
+        if prefetch_batches < 0:
             raise MisconfigurationException("`prefetch_batches` should at least be 0.")
 
         self.prefetch_batches = prefetch_batches + 1
 
         self.dataloader: Optional[Iterable] = None
         self.dataloader_iter: Optional[Iterator] = None
 
+        self.stage: Optional[str]
+        self.batch_to_device: Optional[Callable]
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:72]: Yes, but this is called `batch_to_device` as it needs to move the data on device + any form of transformations.

File: pytorch_lightning/utilities/fetching.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:93]: Once the DataFetcher is exposed through the Trainer, they won't be available at creation time. 

I prefer the setup function as it is more consistent with other objects we have such as LightningModule, DataModule.

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/processors/iterator_batch_processor.py:111]: Done ! However, this entire component will be dropped in next PR, which is why I didn't bother :)

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:29]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:59]: Because it is a hook and should be placed at the top for better readability.

File: pytorch_lightning/utilities/fetching.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:308]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: pytorch_lightning/utilities/fetching.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: tests/utilities/test_fetching.py (hunk 11 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:181]: This test takes 7 sec. I optimized it as much as possible,

File: pytorch_lightning/utilities/fetching.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:64]: Because there is no way to select `InterBatchParallelDataFetcher`.

This is set directly within the test as 

The next PR will clean this.

```py
trainer.data_connector.data_fetcher = InterBatchParallelDataFetcher()
```

File: pytorch_lightning/trainer/trainer.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:928]: follow ups.

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: tests/utilities/test_fetching.py (hunk 11 above)
Comment (awaelchli) [reply tests/utilities/test_fetching.py:177]: I strongly recommend to reconsider this. 7 seconds is an eternity considering we have 1000's of tests. 
If we can't unit test it < 1s, should we only run it locally perhaps?

**Review by awaelchli — APPROVED** [review]
LGTM 
but the added test time is concerning me

File: tests/utilities/test_fetching.py [hunk 20]
```diff
@@ -84,3 +91,149 @@ def test_misconfiguration_error():
 
     iter(fetcher)
     assert fetcher.loader_iters
+
+
+def get_cycles_per_ms() -> float:
+    """
+    Measure and return approximate number of cycles per millisecond for torch.cuda._sleep
+    Copied from: github.com/pytorch/pytorch/blob/master/test/test_cuda.py
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:99]: Maybe i'm blind but I don't see this function in the link, can you fix it?

File: tests/utilities/test_fetching.py [hunk 21]
```diff
@@ -84,3 +91,149 @@ def test_misconfiguration_error():
...
+        return cycles_per_ms
+
+    # Get 10 values and remove the 2 max and 2 min and return the avg.
+    # This is to avoid system disturbance that skew the results, e.g.
+    # the very first cuda call likely does a bunch of init, which takes
+    # much longer than subsequent calls.
+    #
+    # Tested on both Tesla V100, Quadro GP100, Titan RTX, RTX 3090 GPUs
+    # and seems to return stable values. Therefore, we enable caching
+    # using lru_cache decorator above.
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:119]: These comments don't seem to apply

```suggestion
```

File: tests/utilities/test_fetching.py [hunk 22]
```diff
@@ -11,16 +11,23 @@
...
 from pytorch_lightning.trainer.supporters import CombinedLoader
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
-from pytorch_lightning.utilities.fetching import DataFetcher
+from pytorch_lightning.utilities.fetching import DataFetcher, DataLoaderIterDataFetcher, InterBatchParallelDataFetcher
+from tests.helpers.boring_model import BoringModel
+from tests.helpers.runif import RunIf
 
 
-@pytest.mark.parametrize("use_combined_loader", [False, True])
+@pytest.mark.parametrize("use_combined_loader", [False])
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:30]: Is this change on purpose or leftover debugging?

**Review by carmocca — APPROVED** [review]

File: tests/utilities/test_fetching.py (hunk 20 above)
Comment (kaushikb11) [reply tests/utilities/test_fetching.py:99]: https://github.com/pytorch/pytorch/blob/master/test/test_cuda.py#L81

File: tests/utilities/test_fetching.py (hunk 22 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:30]: This is done on purpose. Unittesting with CombinedLoader too.
